For me on Ubuntu 11.10, inotify can emit MOVE_TO events without sending a prior
MOVE_FROM event - which means the "move_src_path" can be None.

To repro I used svn revert of a file, using subversion 1.7 and a file that was
not in the root directory of the checkout - then since the move_from path is
outside of the monitored directories, I didn't receive a move_from event for
the reverting of the file. Without this patch I would fail to see any updates
for a file after reverting it in subversion.

http://bugs.activestate.com/show_bug.cgi?id=91607

Index: /home/toddw/as/komodo-devel/contrib/watchdog/src/watchdog/observers/inotify.py
===================================================================
--- watchdog/src/watchdog/observers/inotify.py	(revision 67502)
+++ watchdog/src/watchdog/observers/inotify.py	(working copy)
@@ -549,9 +549,16 @@
                             del self._wd_for_path[move_src_path]
                             self._wd_for_path[inotify_event.src_path] = moved_wd
                             self._path_for_wd[moved_wd] = inotify_event.src_path
-                        src_path = absolute_path(os.path.join(wd_path, name))
-                        inotify_event = InotifyEvent(wd, mask, cookie, name,
-                                                  src_path)
+                        if move_src_path is None:
+                            # Don't know where it was moved from, so we just say
+                            # it got created instead.
+                            print "Re-mapping to a CREATE: %r %r" % (name, src_path)
+                            inotify_event = InotifyEvent(wd, InotifyConstants.IN_CREATE,
+                                                         0, name, src_path)
+                        else:
+                            src_path = absolute_path(os.path.join(wd_path, name))
+                            inotify_event = InotifyEvent(wd, mask, cookie, name,
+                                                         src_path)
 
                     if inotify_event.is_ignored:
                     # Clean up book-keeping for deleted watches.
